src/core/file.cc

```diff
@@ -1126,15 +1126,17 @@ make_file_impl(int fd, file_open_options options, int flags, struct stat st) noe
                 fsi.fsync_is_exclusive = true;
                 fsi.nowait_works = false;
             }
+            auto st_dev = st.st_dev;
             s_fstype[st_dev] = std::move(fsi);
+            return make_file_impl(fd, std::move(options), flags, std::move(st));
```

Comment:
I tried, but failed, to understand this change, maybe you can help me understand it...
You have a cache s_fstype of fstatfs results. Before this patch, the complicated code below, using the result of the fstatfs to decide how to open the file, was used regardless of whether the fsstatfs results were previously cached or not. After this patch, unless I'm misreading it something (it's really hard to understand the indentation in github), it seems that when the information isn't cached, you cache it, but then don't use what you cached and unconditionlay use the non-append-challenged variant. Why?

Reply:
> Before this patch, the complicated code below, using the result of the fstatfs to decide how to open the file, was used regardless of whether the fsstatfs results were previously cached or not.

No, if the cached result is there it doesn't call fstat but gets the make_ready_future<>() instantly

```c++
    future<> get_fs_info = s_fstype.count(st_dev) ? make_ready_future<>() :
        engine().fstatfs(fd).then([fd, st_dev] (struct statfs sfs) {
            ...
            s_fstype[st_dev] = std::move(fsi);
        });
```

> After this patch, unless I'm misreading it something (it's really hard to understand the indentation in github), it seems that when the information isn't cached, you cache it, but then don't use what you cached and unconditionlay use the non-append-challenged variant.

I do exactly the same, but without make_ready_future<>. If the result is not there I do

```c++
    if (s_fstype.count(st_dev) == 0) {
        return engine().fstatfs(fd).then([fd, options = std::move(options), flags, st = std::move(st)] (struct statfs sfs) {
            ...
            return ... // "recursive" call to the current function
        });
```

and if the result is not there it just returns the future with file impl _without_ .then()-ing it to the make_ready_future<>()

```c++
    return make_ready_future<shared_ptr<file_impl>>(...);
```

Reply:
Oh, I missed the **recursive** call to the current function. I misread it as a call to another function.

Reply:
Nitpick (not important): I don't think moving "struct stat" actually does anything.